### PR TITLE
[WIP] kubeadm: Carry over new component config versions

### PR DESCRIPTION
**Pull request by rosti**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

Although group based, kubeadm's component configs support is still fixed to
support only a single version. Hence, it doesn't matter if the loaded component
config version is older or newer - kubeadm would always bail out with an error.

Forcing users to the latest version of a component config format would also
mean that for many of them a manual migration would be required (while the old
version is still loadable by the component in question).

Hence, if kubeadm does not require any new features of the config, it does not
need to force users to update their config.

With that in mind, many users may actually want to provide configs in the new
format. Hence, to enable that without bailing out with an error, kubeadm needs
to tolerate newer component config versions without explicitly understanding
that format.

This change allows for that to happen by treating newer config versions as
blobs of YAML. As a consequence, newer configs are not defaulted (patched with
kubeadm preferred values) and are not guaranteed to work at all.

It is a requirement to the user to verify that correct config settings are
supplied or the component in question may not work.

**Which issue(s) this PR fixes**:

Refs:
- kubernetes/kubeadm#1940
- kubernetes/enhancements#1381

**Special notes for your reviewer**:

This PR is part of the implementation of the new kubeadm component config management scheme KEP (see link below).

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/priority important-longterm
/assign @timothysc @fabriziopandini @neolit123 @ereslibre 

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm now allows for component configs in newer unsupported by it formats to be used by treating them as YAML blobs
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md
```

### CODE REVIEW

**Comment by neolit123** [comment]
/retest
@rosti this PR seems good, if you change the release note and WIP it seems mergeble?

**Comment by neolit123** [comment]
> PR is marked as a WIP as it needs some more comments here and there and possibly a few more test cases.

i don't think i have further comments, but keep the WIP if you'd like to add more unit tests.

File: cmd/kubeadm/app/componentconfigs/configset.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/componentconfigs/configset.go]: Just for confirmation.
This approach works for well-known component configs only; whenever addons will start to add unknown component config to the picture, an equivalent solution should be provided

File: cmd/kubeadm/app/componentconfigs/kubelet.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/componentconfigs/kubelet.go]: If I got it right when a new version is created, the version should be hard-coded here, right?
might worth to add a comment?

**Review by fabriziopandini — APPROVED** [review]
@rosti thanks!
LGMT, with only two questions to check if have properly understood the change.

Overall, when I think to the story about component configs being completed, I think it is time to start thinking about how to document this from a user perspective, and how to implements E2E test for ensuring that this set of kubeadm kubeadm feature has no regressions.
Is it possible to open an issue tracking documentation and one for e2e test?

About the doc, it would be great to have something organized around user stories, something like:
- _as an operator I would like to continue to use the current release of the kubelet (kube-proxy) component config until it is supported, no matter a new release exists_
- _as an operator I would like to start using the new release of the kubelet (kube-proxy) component config_

Then, if doable, E2E tests should validate uses stories

**Comment by neolit123** [comment]
@fabriziopandini 

> Then, if doable, E2E tests should validate uses stories

i think we can get away without adding e2e tests for this and only having unit tests.
but a big +1 for docs - @rosti

File: cmd/kubeadm/app/componentconfigs/configset.go
Comment (rosti) [reply cmd/kubeadm/app/componentconfigs/configset.go]: That's quite unlikely. The idea is to have here the component configs of things, that are managed by kubeadm. The configs of things managed by the addon manager can be piped directly to it without ever going through this module.
In fact configs (such as the kube-proxy one), can be optionally excluded at some point from this list if an addon manager is used.

File: cmd/kubeadm/app/componentconfigs/kubelet.go
Comment (rosti) [reply cmd/kubeadm/app/componentconfigs/kubelet.go]: Well, in general that's not the only change, that's going to happen. Besides that, the developer who updates it needs to verify that the correct types are used in the `kubeletConfig` struct and that everything is marshaled accordingly.
So, in general, the whole `kubelet*.go` needs to be checked to see what needs bumping and what not.

**Comment by rosti** [comment]
Reverted back to WIP here. The Instance specific vs Shared component configs KEP (as proposed by kubernetes/enhancements#1439) might cause issues with the current implementation. Hence it's safer to hold this for now until the KEP merges and we can have a clear view on how to proceed with such component configs.

**Pull request closed without merging**
